### PR TITLE
Rewrite to use discovery-swarm-web and random-access-web

### PR DESCRIPTION
**Pull request by RangerMauve**
I finally got to put some of the high level pieces together in a new Dat-JS update, and got the unit tests working! 😀

The coolest part is that the discovery server acts as a proxy for discovery-swarm, so it can't snoop on what you're pushing through it, as well as enabling a wider range of applications that can make use of it.

Plus, it will attempt to load a local gateway before reaching out to the internet, so we have a path towards some sort of daemon to put on people's machines to enable this use case (plus pinning?).

Added:
- [discovery-swarm-web](https://github.com/RangerMauve/discovery-swarm-web) support
- New `persist` flag that will automatically set [random-access-web](https://github.com/RangerMauve/random-access-web) as the storage layer
- Everything works out of the box with the default gateway (hopefully I won't regret this)

Removed:
- Got rid of the `repo` abstraction and used hyperdrives directly

TODO:

- [x] I'm not a fan of the random-access-web API so I'm going to rework it to not need promises.
- [ ] @tinchoz49 is working on an update to [discovery-swarm-webrtc](https://github.com/geut/discovery-swarm-webrtc#readme) to improve performance. I'll try to help out, but I don't think it's blocking for this release.

### CODE REVIEW

**Comment by okdistribute** [comment]
This looks amazing!! I'll have more time to review either tomorrow or next week, but don't let that stop you from merging if you don't get any substantive comments over the next few days. Cheers & congrats!
Cc @substack

File: package.json [hunk 1]
```diff
@@ -18,23 +19,19 @@
...
-    "pump": "^1.0.3",
     "random-access-memory": "^2.4.0",
-    "run-parallel": "^1.1.9",
-    "signalhubws": "^1.0.10",
-    "sodium-universal": "^2.0.0",
-    "through2": "^2.0.0",
-    "webrtc-swarm": "^2.9.0",
-    "websocket-stream": "^5.1.2",
-    "xtend": "^4.0.1"
+    "random-access-web": "github:random-access-storage/random-access-web",
```
Comment (joehand) [inline package.json]: Can we get this published before merging this PR?

File: index.js [hunk 2]
```diff
@@ -1,85 +1,157 @@
...
-    return !!this.repos.find((repo) => repo.url === url)
+    return !!this.archives.find((archive) => archive.url === url)
+  }
+
+  _replicate (info) {
+    var stream = hypercoreProtocol({
+      id: this.opts.id,
+      live: true,
+      encrypt: true,
+      extensions: ['ephemeral', 'session-data']
```
Comment (joehand) [inline index.js]: Can you give me a tldr on how we're using these extensions? (not sure how many people are using these yet so could be a good example use case)

File: README.md [hunk 3]
```diff
@@ -32,67 +32,47 @@ var readStream = repo.archive.readFile('hello.txt', function (err, data) {
...
 ```js
 var Dat = require('dat-js')
-var db = require('random-access-idb')('dats')
 
 var dat = new Dat()
 
-var repo = dat.create({
-  db: db
+var archive = dat.create({
+  persist: true
```
Comment (joehand) [inline README.md:42]: Can you add a sentence or two on how this works and what happens if you don't use this? (i.e. link to the repos and say what "persist" means in terms of browser storage.

Comment (joehand) [reply README.md:42]: Oh I see this in in API section. May be good to mentio here too since it is a whole section on it =)

File: README.md [hunk 4]
```diff
@@ -106,13 +86,14 @@ repo.archive.readFile('/example.txt', 'utf-8', (err, data) => {
...
-  * `gateway`: An optional string or array of strings for [dat-gateway](https://github.com/garbados/dat-gateway/) instances for websocket replication. This is used to proxy data from the rest of the dat network when there are no WebRTC peers available with your data. There is no gateway set by default, but you can use `https://gateway.mauve.moe/` if you want it.
+ * `options`: any options you can pass to [mafintosh/hyperdrive](https://github.com/mafintosh/hyperdrive). These options will become default for all dats. It also gets passed as options into [discovery-swarm-web](https://github.com/RangerMauve/discovery-swarm-web). In addition it has the following:
+  * `persist`: Whether the data should persist locally or load in memory. Default: `false` (memory only). This uses [random-access-web](https://github.com/RangerMauve/random-access-web) for persistence to choose the best storage layer for the current browser.
+  * `db`: Pass in the random-access-storage](https://github.com/random-access-storage/random-access-storage) instance to use. (overrides the `persist` option)
```
Comment (joehand) [inline README.md]: fix link, missing first `[`

File: README.md [hunk 5]
```diff
@@ -106,13 +86,14 @@ repo.archive.readFile('/example.txt', 'utf-8', (err, data) => {
...
+  * `id`: The ID to use when replicating hyperdrives
```
Comment (joehand) [inline README.md:70]: What is this for?

File: README.md [hunk 6]
```diff
@@ -106,13 +86,14 @@ repo.archive.readFile('/example.txt', 'utf-8', (err, data) => {
...
 
 ### `dat.get(url, [options])`
 
-Adds a new dat with the given url. Joins the appropriate swarm for that url and begins to upload and download data. If the dat was already added, it will return the existing instance. One gotcha is that dat-js doesn't support DNS resolution yet. As such you'll need to use the actual archive key for loading websites.
+Adds a new dat with the given url. Joins the appropriate swarm for that url and begins to upload and download data. If the dat was already added, it will return the existing instance. One gotcha is that dat-js doesn't support DNS resolution yet. As such you'll need to use the actual archive key for loading websites. `dat-js` adds a `url` field to the archive, but you can see the rest of the APIs available in the [hyperdrive](https://www.npmjs.com/package/hyperdrive) docs.
 
  * `url`: Either a `dat://` url or just the public key in string form.
```
Comment (joehand) [inline README.md:76]: Been trying to update `public key` -> `read key`. Not sure what to do with url though... (need to really get that nomeclature draft in)

File: index.js [hunk 7]
```diff
@@ -1,85 +1,157 @@
...
+    const discoveryOpts = Object.assign({
+      stream: (info) => this._replicate(info)
+    }, this.opts)
+    this.swarm = new DiscoverySwarmWeb(discoveryOpts)
+
+    const storageOpts = Object.assign({
+      name: STORAGE_NAME,
+      storeName: STORAGE_COLLECTION_NAME
+    }, this.opts)
+    this.persistence = RAW(storageOpts)
```
Comment (joehand) [inline index.js:40]: What options are being passed on? RAW doesn't have any options documented, so not clear now these are being used.

File: README.md [hunk 8]
```diff
@@ -32,67 +32,47 @@ var readStream = repo.archive.readFile('hello.txt', function (err, data) {
...
-  db: db
+var archive = dat.create({
+  persist: true
 })
 
-repo.archive.writeFile('/example.txt', 'Hello World!', () => {
+archive.writeFile('/example.txt', 'Hello World!', () => {
   // Save it for later
-  localStorage.setItem('My_Repo', repo.url)
+  localStorage.setItem('My_Repo', archive.url)
```
Comment (joehand) [inline README.md:47]: I guess this + the random-access-web options are confusing me. If it is persisting then why do you have to also manually persist it? We should specify *what* is being persisted.

**Review by joehand — COMMENTED** [review]
Sweet! 

Left some comments. I'd like to try updating my test web project but feel free to merge before, not sure when I'll have a chance to do that.

File: index.js (hunk 2 above)
Comment (RangerMauve) [reply index.js]: They're not being used, I copied this part from beaker. I want to eventually support them, though. I'll remove it for now.

File: package.json (hunk 1 above)
Comment (RangerMauve) [reply package.json]: Whoops! I forgot to update the dependency

File: README.md (hunk 5 above)
Comment (RangerMauve) [reply README.md:70]: Not sure! Modules related to webrtc peer discovery use an ID. https://github.com/geut/discovery-swarm-webrtc/blob/sub-signalhub/index.js#L40

File: index.js (hunk 7 above)
Comment (RangerMauve) [reply index.js:40]: I'll add to the docs, but `name` is the only one that does something across different storage layers.

File: README.md (hunk 8 above)
Comment (RangerMauve) [reply README.md:47]: By default it's all in memory, I have the `persist` flag there for when you want to load archives into and from persistent storage. All the data in the archive would be getting persisted.

File: README.md (hunk 6 above)
Comment (RangerMauve) [reply README.md:76]: I added a thing to mention that the `url` is the `read key`

File: index.js (hunk 7 above)
Comment (RangerMauve) [reply index.js:40]: I added docs on the options: https://www.npmjs.com/package/random-access-web#options

File: README.md (hunk 3 above)
Comment (RangerMauve) [reply README.md:42]: Added a comment relating to it. Should I have a description instead?

File: README.md (hunk 5 above)
Comment (joehand) [reply README.md:70]: haha ok. I think all the [discovery modules](https://github.com/mafintosh/discovery-swarm/blob/master/index.js#L39) have a peer id but I guess never seen it out as an option . So thats fine i guess.

File: README.md (hunk 8 above)
Comment (joehand) [reply README.md:47]: But you still need to remember/save the archive key somewhere, right? That should be explicitly discussed.

Comment (RangerMauve) [reply README.md:47]: That's a good point. I'll add a comment to the create example and to the API docs

Comment (RangerMauve) [reply README.md:47]: That's a great point though. Why *should* they need to save the URL somewhere? It'd be nice if they could have named archives the way [multifeed](https://github.com/kappa-db/multifeed) does. 🤔 Probably for the next update, though.

**Comment by RangerMauve** [comment]
I'm testing this with [load-dat-page](https://github.com/RangerMauve/load-dat-page) and I'm having some trouble getting stuff to load. The first piece of information loads fine, but after that it seems like calls to hyperdrive lock up. 😕

**Comment by RangerMauve** [comment]
Whoops! I wasn't detecting if an archive was already loaded or not properly. :)

**Comment by joehand** [comment]
I tried to update my datpublish thing to use this and had some issues just trying to use it based on current documentation... 

But I'll check out the load-dat-page and see if I was doing something obviously wrong.

**Comment by RangerMauve** [comment]
Oh dang. Mind linking me to the branch that's failing?

**Comment by joehand** [comment]
Got it working last night. It seems like something with the [attempted localhost connection](https://github.com/RangerMauve/discovery-swarm-web/blob/master/index.js#L145) is the cause. If I replaced localhost url with discovery url it worked fine. 

(code on my other computer and I keep forgetting to push...)

**Comment by RangerMauve** [comment]
@joehand Try reinstalling discovery-swarm-web. I fixed [a bug](https://github.com/RangerMauve/discovery-swarm-web/commit/580aa691fd4caac3fef4c972bbbbb6e16013a9bd) relating to localhost in v1.0.5

**Comment by RangerMauve** [comment]
KK, I'm releasing it. :D

**Comment by tinchoz49** [comment]
awesome work @RangerMauve !

**Comment by RangerMauve** [comment]
It's published in [v8.0.0](https://github.com/datproject/dat-js/releases/tag/v8.0.0)

**Comment by okdistribute** [comment]
Looks great!

**Comment by cblgh** [comment]
woah this is big, nicely done @RangerMauve!

**Pull request merged**